File: backend/meta/relation.py
"""
关系实体
Relation连接两个Model，不依赖Domain
"""
from typing import Optional


class Relation:
    """关系实体 - 连接两个Model"""
    
    def __init__(
        self,
        id: int,
        name: str,
        sourceModelId: int,
        targetModelId: int,
        type: str = "one-to-many",
        description: Optional[str] = None,
        enabled: bool = True
    ):
        self.id = id
        self.name = name
        self.sourceModelId = sourceModelId
        self.targetModelId = targetModelId
        self.type = type
        self.description = description
        self.enabled = enabled
    
    def to_dict(self) -> dict:
        """转换为字典"""
        return {
            "id": self.id,
            "name": self.name,
            "sourceModelId": self.sourceModelId,
            "targetModelId": self.targetModelId,
            "type": self.type,
            "description": self.description,
            "enabled": self.enabled
        }
    
    @classmethod
    def from_dict(cls, data: dict) -> 'Relation':
        """从字典创建实例"""
        return cls(
            id=data["id"],
            name=data["name"],
            sourceModelId=data["sourceModelId"],
            targetModelId=data["targetModelId"],
            type=data.get("type", "one-to-many"),
            description=data.get("description"),
            enabled=data.get("enabled", True)
        )
    
    def toggle(self) -> 'Relation':
        """切换启用状态"""
        self.enabled = not self.enabled
        return self
    
    def __repr__(self):
        return f"Relation(id={self.id}, name='{self.name}', source={self.sourceModelId}, target={self.targetModelId})"

File: backend_java/src/main/java/com/jianmo/platform/entity/Relation.java
package com.jianmo.platform.entity;

import javax.persistence.*;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.AllArgsConstructor;
import lombok.ToString;

import java.time.LocalDateTime;


@Entity
@Table(name = "relations")
@Data
@NoArgsConstructor
@AllArgsConstructor
public class Relation {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    private String name;

    @Column(name = "source_model_id")
    private Long sourceModelId;

    @Column(name = "target_model_id")
    private Long targetModelId;

    private String type = "one-to-many";

    @Column(columnDefinition = "TEXT")
    private String description;

    @Column(name = "is_enabled")
    private Boolean isEnabled = true;

    @Column(name = "created_at")
    private LocalDateTime createdAt;

    @Column(name = "updated_at")
    private LocalDateTime updatedAt;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "source_model_id", insertable = false, updatable = false)
    @ToString.Exclude
    private Model sourceModel;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "target_model_id", insertable = false, updatable = false)
    @ToString.Exclude
    private Model targetModel;

    
    protected void onCreate() {
        createdAt = LocalDateTime.now();
        updatedAt = LocalDateTime.now();
    }

    
    protected void onUpdate() {
        updatedAt = LocalDateTime.now();
    }

    public Relation(String name, Long sourceModelId, Long targetModelId, String type) {
        this.name = name;
        this.sourceModelId = sourceModelId;
        this.targetModelId = targetModelId;
        this.type = type;
    }

    public boolean involvesModel(Long modelId) {
        return (sourceModelId != null && sourceModelId.equals(modelId)) ||
               (targetModelId != null && targetModelId.equals(modelId));
    }

    public boolean isValid() {
        return sourceModelId != null && targetModelId != null && !sourceModelId.equals(targetModelId);
    }

    public Boolean getEnabled() { return isEnabled; }
    public void setEnabled(Boolean enabled) { isEnabled = enabled; }
}
